core/templates/dev/head/services/ComputeGraphService.ts

```diff
@@ -17,74 +17,80 @@
  * exploration.
  */
 
-angular.module('oppia').factory('ComputeGraphService', [
-  function() {
-    var _computeGraphData = function(initStateId, states) {
-      var nodes = {};
-      var links = [];
-      var finalStateIds = states.getFinalStateNames();
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
 
-      states.getStateNames().forEach(function(stateName) {
-        var interaction = states.getState(stateName).interaction;
-        nodes[stateName] = stateName;
-        if (interaction.id) {
-          var groups = interaction.answerGroups;
-          for (var h = 0; h < groups.length; h++) {
-            links.push({
-              source: stateName,
-              target: groups[h].outcome.dest,
-            });
-          }
+@Injectable({
+  providedIn: 'root'
+})
+export class ComputeGraphService {
+  _computeGraphData(initStateId: any, states: any): any {
```

Comment:
Is this really `any` or you didn't spend time in finding the type? If its the latter one, it's totally fine to me :), but please do link it to the required TODO.

Reply:
@bansalnitish Well actually i tried to find out the exact datatypes, but there are some complex things going on here, i actually also tried some calculated guesses but it failed. So i guess any is best suited here. But if you suggest, i can add a // todo. So what do you suggest? @bansalnitish

Reply:
Let's add a todo. We will take a more deeper look when resolving the TODO's.

Reply:
Done

Reply:
I'm going to push back on this. We should not be introducing TODOs so easily. Please figure out what the types are and fix them in this PR.

Reply:
This also applies as a general rule for all other similar cases in other PRs.

If a type is hard to figure out, there needs to be (at minimum) a full comment and justification explaining what has been tried and what the issue is. But we should avoid introducing more TODOs and instead try to do things correctly the first time.